password/src/Data/Password/Validate.hs

```diff
@@ -0,0 +1,243 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+{-|
+Module      : Data.Password.Valid
+Copyright   : (c) Dennis Gosnell, 2019; Felix Paulusma, 2020; Hiroto Shioi, 2020
+License     : BSD-style (see LICENSE file)
+Maintainer  : cdep.illabout@gmail.com
+Stability   : experimental
+Portability : POSIX
+
+= validate
+
+It is common for passwords to have a set of requirements.
+(e.g. Password must contain at least eight characters that consist of alphabetic
+characters combined with numbers or special characters.)
+
+Validate module provides set of functions which enables you to validate them.
+
+-}
+
+module Data.Password.Validate
+  ( -- * Data type
+    PasswordPolicy (..),
+    InvalidReason (..),
+    CharacterCategory(..),
+    -- * Default
+    defaultPasswordPolicy,
+    -- * Functions
+    valid,
+    validatePassword,
+    -- * Utility
+    isValidPasswordPolicy,
+    defaultCharSet,
+  ) where
+
+import           Data.Char              (isDigit, isLower, isUpper)
+import           Data.Maybe             (catMaybes, fromMaybe)
+import           Data.Password.Internal (Password (..))
+#if! MIN_VERSION_base(4,13,0)
+import           Data.Semigroup         ((<>))
+#endif
+import           Data.Text              (Text)
+import qualified Data.Text              as T
+import           Test.Tasty.QuickCheck  (Arbitrary (..), Gen, choose, elements,
+                                         oneof)
+
+-- $setup
+-- >>> :set -XOverloadedStrings
+--
+-- Import needed libraries.
+--
+-- >>> import Data.Password
+
+-- | Set of policies used to validate 'Password'
+data PasswordPolicy = PasswordPolicy
+    { passPolicyMinimumLength :: !Int
+    -- ^ Required password minimum length
+    , passPolicyMaximumLength :: !Int
+    -- ^ Required password maximum length
+    , passPolicyCharUppercase :: !(Maybe Int)
+    -- ^ Required number of upper-case characters
+    , passPolicyCharLowercase :: !(Maybe Int)
+    -- ^ Required number of lower-case characters
+    , passPolicyCharSpecial   :: !(Maybe Int)
+    -- ^ Required number of special characters
+    , passPolicyCharDigit     :: !(Maybe Int)
+    -- ^ Required number of ASCII-digit characters
+    , passPolicyCharSet       :: !Text
+    -- ^ Set of characters that can be used for password
+    }
+    deriving (Eq, Ord, Show)
+
+-- | Generate valid PasswordPolicy
+instance Arbitrary PasswordPolicy where
+  arbitrary = do
+    minimumLength <- choose (1, 10)
+    upperCase <- genMaybeInt
+    lowerCase <- genMaybeInt
+    special <- genMaybeInt
+    digit <- genMaybeInt
+    let sumLength = sum $ catMaybes [upperCase, lowerCase, special, digit]
+    let minMaxLength = maximum [minimumLength, sumLength]
+    maximumLength <- choose (minMaxLength, minMaxLength + 10)
+    return $ PasswordPolicy minimumLength maximumLength upperCase lowerCase special digit defaultCharSet
+    where
+      genMaybeInt :: Gen (Maybe Int)
+      genMaybeInt = oneof [return Nothing, Just <$> (choose (1, 10))]
```

Comment:
Can be shorter with `liftArbitrary (choose (1, 10))` from the `Arbitrary1` class. Unless you want it to be `Nothing` 50% of the time. (`Arbitrary1` instance of `Maybe` uses `Nothing` 25% of the time)